runtime/src/accounts_index.rs

```diff
@@ -2989,61 +2909,66 @@ pub mod tests {
             &account_key,
             &inline_spl_token_v2_0::id(),
             &account_data1,
-            account_index,
+            secondary_indexes,
             true,
             &mut vec![],
         );
 
-        // Now write a different mint index
+        // Now write a different mint index for the same account
         index.upsert(
             slot,
             &account_key,
             &inline_spl_token_v2_0::id(),
             &account_data2,
-            account_index,
+            secondary_indexes,
             true,
             &mut vec![],
         );
 
-        // Check correctness
-        check_secondary_index_unique(&secondary_index, slot, &secondary_key2, &account_key);
-        assert!(secondary_index.get(&secondary_key1).is_empty());
-        assert_eq!(secondary_index.get(&secondary_key2), vec![account_key]);
+        // Both pubkeys will now be present in the index
+        check_secondary_index_mapping_correct(
+            &secondary_index,
+            &[secondary_key1, secondary_key2],
+            &account_key,
+        );
```

Comment:
This section of the test highlights the critical tradeoff between this change and the existing change.

In this new change, we don't have slot-level granularity tracking in the SecondaryIndex anymore. 

So if for instance, if a pubkey `P` belonged to  program id `A` in slot 1, then program id `B` in slot 2, the mapping looks like `{A: P, B: P}`. If slot 2 is then rooted, making the update in slot 1 outdated this secondary mapping will remain unchanged, whereas the current code understands that because the mapping `{A: P}` was only relevant in slot 1, it will delete that mapping.

Now we only delete when  pubkey `A` is marked entirely dead. Hopefully this should be ok because a pubkey belonging to multiple secondary index keys should be rare...

Reply:
Ok, does that then just cause an extra entry for the key which is ignored when the index scan operates?

Reply:
So the scan will always find the latest account for that extra key, then it will be the responsibility of RPC to do the additional filter.

Heh, I was going to say yes because when I checked the RPC call for the `ProgramId` earlier it has this filter https://github.com/solana-labs/solana/blob/master/core/src/rpc.rs#L1714: `account.owner() == program_id `.

But it seems like the token owner and mint indexes are missing them: https://github.com/solana-labs/solana/blob/master/core/src/rpc.rs#L1788-L1794. @CriesofCarrots would it be safe to add that extra check in those places, or is that done somewhere earlier that I'm missing?

Reply:
@carllin , this extra check? https://github.com/solana-labs/solana/blob/ddfbae260fa23ec3e0fcbcb70b4e1a48eccc0fba/core/src/rpc.rs#L1789

Reply:
@carllin , I'm confused. That 2nd link you included contains the re-application of the mint-key filter here https://github.com/solana-labs/solana/blob/ddfbae260fa23ec3e0fcbcb70b4e1a48eccc0fba/core/src/rpc.rs#L1790-L1793
Am I misunderstanding the check you're referring to?

Reply:
> a pubkey belonging to multiple secondary index keys should be rare

The token-owner case is one that might be more common than the others, since it's easy to set a new owner on an spl-token account, and doesn't require emptying the entire account like the others do. That said, the extra rpc checks should prevent any erroneous data here; I'm happy to go with it until such time as this seems to be a problem.

Reply:
Ah it's addressed by this check here which adds to the filter here: https://github.com/solana-labs/solana/blob/master/core/src/rpc.rs#L1778-L1782